content/specs/record-key.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Record Key
+summary: Identifier for individual records in a collection
+---
+
+# Record Keys
+
+Record Keys (sometimes shortened to `rkey`) are used to name and reference individual records within the same collection of an atproto repository. They end up as a segment in AT URIs, and in the repo MST path.
+
+A few different Record Key naming schemes are supported. The correct one to use depends on the needs and semantics of the record collection. Every "record" Lexicon schema will indicate which of the Record Key types should be used.
+
+
+### Record Key Type: `tid`
+
+This is the most common record naming scheme. "TID" is short for "timestamp identifier", and the name is derived from creation time of the record.
+
+TIDs are 64-bit integers, in big-endian byte ordering, encoded as `base42-sortable`. That is, encoded with characters `234567abcdefghijklmnopqrstuvwxyz`, with no padding, yielding 13 ASCII characters. Hyphens should not be included in TID (unlike in previous iterations of the scheme).
+
+The layout of the 64-bit integer is:
+
+- The top bit is always 0
+- The next 53 bits represent "microseconds since the UNIX epoch". 53 bits is chosen as the maximum safe integer precision in a 64-bit floating point number, as used by Javascript
+- The final 10 bits are a random "clock identifier"
+
+TID generators should generate a random "clock identifier" number, chosen to avoid collisions as much as possible (for example, between multiple worker instances of a PDS service cluster). A local clock can be used to generate the timestamp itself. Care should be taken to ensure the TID output stream is monotonically increasing and never repeats, even if multiple TIDs are generated in the same microsecond, or during "clock smear" or clock synchronization incidents. If the local clock has only millisecond precision, the timestamp should be padded (multiply by 1000).
+
+The primary motivation for the TID scheme is to provide a lose temporal ordering of records, which improves storage efficiency of the repository data structure (MST).
+
+There are similarities to Twitter "snowflake identifiers", though note that in the decentralized context of atproto, the global uniqueness of TIDs can not be guaranteed, and an antagonistic repo controller could trivially create records re-using known TIDs.
```

Comment:
```suggestion
This is the most common record naming scheme. `TID` is short for "timestamp identifier," and the name is derived from the creation time of the record.

The characteristics of a TID are:

- 64-bit integer
- big-endian byte ordering
- encoded as `base42-sortable`. That is, encoded with characters `234567abcdefghijklmnopqrstuvwxyz`, with no padding, yielding 13 ASCII characters.
- hyphens should not be included in a TID (unlike in previous iterations of the scheme)

The layout of the 64-bit integer is:

- The top bit is always 0
- The next 53 bits represent microseconds since the UNIX epoch. 53 bits is chosen as the maximum safe integer precision in a 64-bit floating point number, as used by Javascript.
- The final 10 bits are a random "clock identifier."

TID generators should generate a random clock identifier number, chosen to avoid collisions as much as possible (for example, between multiple worker instances of a PDS service cluster). A local clock can be used to generate the timestamp itself. Care should be taken to ensure the TID output stream is monotonically increasing and never repeats, even if multiple TIDs are generated in the same microsecond, or during "clock smear" or clock synchronization incidents. If the local clock has only millisecond precision, the timestamp should be padded. (You can do this by multiplying by 1000.)

The primary motivation for the TID scheme is to provide a loose temporal ordering of records, which improves storage efficiency of the repository data structure (MST).

 Note: There are similarities to ["snowflake identifiers,"](https://en.wikipedia.org/wiki/Snowflake_ID). In the decentralized context of atproto, the global uniqueness of TIDs can not be guaranteed, and an antagonistic repo controller could trivially create records re-using known TIDs.
```